### ISSUE TITLE
[BUG] Most uses of `ObjectExt::set_qdata` leak memory

### ISSUE THREAD

**Issue by mahkoh** [gtk-rs/gtk4-rs#564]
In the following example I'm subclassing `gtk4::EntryBuffer` in order for `gtk_entry_buffer_get_text` to call into gtk4-rs. In the trampoline, `set_qdata` is called to keep the returned data alive:

https://github.com/gtk-rs/gtk4-rs/blob/fb59005041d06bc275de3ff23a740395b6332c1e/gtk4/src/subclass/entry_buffer.rs#L234-L238

`set_qdata` is called with a pointer returned by the equivalent of `Box::leak`. However, `set_qdata` expects to be passed a type with a destructor that will clean up owned resources:

https://github.com/gtk-rs/gtk-rs-core/blob/9ea03f8eaacd84aa1cc850765748a779757ce93a/glib/src/object.rs#L1641-L1655

Callers of `set_qdata` should therefore store a pointer to the target value and then pass the box itself to `set_qdata`.

Almost all users of `set_qdata` use the same incorrect pattern.

Before running the following example, note that it will rapidly allocate memory.

```rust
use imp::EbImp;
use gtk4::subclass::prelude::*;
use gtk4::prelude::*;
use gtk4::glib::Object;

mod imp {
    #[derive(Default)]
    pub struct EbImp;
}

#[gtk4::glib::object_subclass]
impl ObjectSubclass for EbImp {
    const NAME: &'static str = "EbImp";
    type Type = Eb;
    type ParentType = gtk4::EntryBuffer;
    type Interfaces = ();
}

glib::wrapper! {
    pub struct Eb(ObjectSubclass<EbImp>) @extends gtk4::EntryBuffer;
}

impl ObjectImpl for EbImp {}

impl EntryBufferImpl for EbImp {}

fn main() {
    gtk4::init().unwrap();

    let entry_buffer: Eb = Object::new(&[]).unwrap();
    entry_buffer.set_text(&"a".repeat(4096));
    loop {
        entry_buffer.text();
    }
}
```

**Comment by sdroege** [comment]
Yes all these cases need to store a struct that correctly frees the pointer on `Drop::drop`. Storing raw pointers is only valid if they must not be freed.

Do you want to submit a PR for this for gtk4-rs? I'm going to fix gtk-rs-core now.

**Comment by sdroege** [comment]
See https://github.com/gtk-rs/gtk-rs-core/pull/257 for the gtk-rs-core one.
